File: apps/web/lib/simulation/engine.ts
import {
  actionDefinitions,
  evidenceCatalog,
  MAX_LOGS,
  MAX_METRICS,
  SIMULATION_INTERVAL_SECONDS,
  topology,
} from "./scenario";
import type {
  ActionId,
  Health,
  IncidentStage,
  LogEntry,
  MetricPoint,
  ServiceState,
  SimulationEvent,
  SimulationState,
  TimelineEntry,
} from "./types";

function hash(text: string): number {
  let value = 2166136261;
  for (const char of text) {
    value ^= char.charCodeAt(0);
    value = Math.imul(value, 16777619);
  }
  return value >>> 0;
}
export function seedFromSession(scenarioId: string, sessionId: string): number {
  return hash(`${scenarioId}:${sessionId}`);
}
function noise(seed: number, tick: number, salt: number): number {
  let x = seed ^ Math.imul(tick + 1, 0x9e3779b1) ^ salt;
  x ^= x << 13;
  x ^= x >>> 17;
  x ^= x << 5;
  return (x >>> 0) / 4294967295;
}

export function stageFor(
  second: number,
  mitigationAt: number | null,
): IncidentStage {
  if (mitigationAt !== null) {
    const since = second - mitigationAt;
    if (since >= 120) return "Completed";
    if (since >= 60) return "Recovery";
    return "Incident mitigation";
  }
  if (second >= 150) return "Checkout errors";
  if (second >= 120) return "Order-service latency increase";
  if (second >= 90) return "Database pool saturation";
  if (second >= 60) return "Connection leak begins";
  if (second >= 30) return "Deployment completed";
  return "Normal";
}

function severity(stage: IncidentStage): number {
  return Math.max(
    0,
    [
      "Normal",
      "Deployment completed",
      "Connection leak begins",
      "Database pool saturation",
      "Order-service latency increase",
      "Checkout errors",
    ].indexOf(stage),
  );
}
function servicesFor(
  state: Pick<
    SimulationState,
    "stage" | "elapsedSeconds" | "seed" | "tick" | "modifiers"
  >,
): ServiceState[] {
  const level = severity(state.stage);
  const recovering =
    state.stage === "Incident mitigation" || state.stage === "Recovery";
  const completed = state.stage === "Completed";
  return topology.map((service, index) => {
    let health: Health = "healthy";
    if (service.id === "order" && level >= 3)
      health = level >= 5 ? "critical" : "degraded";
    if (service.id === "orders-db" && level >= 3)
      health = level >= 4 ? "critical" : "degraded";
    if (["web", "gateway", "cart"].includes(service.id) && level >= 5)
      health = "degraded";
    if (recovering && ["order", "orders-db", "gateway"].includes(service.id))
      health = "recovering";
    if (completed) health = "healthy";
    const jitter = Math.round(noise(state.seed, state.tick, index + 1) * 8);
    const affected = ["order", "orders-db"].includes(service.id);
    const latency = completed
      ? 90 + jitter
      : affected
        ? 90 + level * 470 - state.modifiers.latencyReduction + jitter
        : 55 + (level >= 5 ? 120 : 0) + jitter;
    const errors = completed
      ? 0.2
      : service.id === "order"
        ? Math.max(0.2, level * 2.7 - state.modifiers.errorReduction)
        : ["web", "gateway", "cart"].includes(service.id) && level >= 5
          ? 4.5
          : 0.2;
    return {
      ...service,
      dependencies: [...service.dependencies],
      health,
      requestsPerMinute: 780 + jitter * 3,
      errorRate: Number(errors.toFixed(1)),
      latencyMs: Math.max(20, Math.round(latency)),
    };
  });
}

function metricFor(state: SimulationState): MetricPoint {
  const level = severity(state.stage);
  const recoveryFactor =
    state.stage === "Recovery" ? 0.35 : state.stage === "Completed" ? 0 : 1;
  const jitter = Math.round(noise(state.seed, state.tick, 99) * 22);
  return {
    second: state.elapsedSeconds,
    orderLatencyMs: Math.max(
      80,
      Math.round(
        (95 + level * 480 - state.modifiers.latencyReduction) * recoveryFactor +
          jitter,
      ),
    ),
    checkoutErrorRate: Number(
      Math.max(
        0.1,
        (level >= 5
          ? 12.8 - state.modifiers.errorReduction
          : level >= 4
            ? 2.2
            : 0.2) * recoveryFactor,
      ).toFixed(1),
    ),
    dbPoolUsed: Math.min(
      40 + state.modifiers.poolBonus,
      Math.round((8 + level * 7) * recoveryFactor),
    ),
    dbPoolMax: 40 + state.modifiers.poolBonus,
  };
}
function logFor(state: SimulationState): LogEntry {
  const level = severity(state.stage);
  const messages =
    level >= 5
      ? ["ERROR", "checkout request failed after upstream timeout"]
      : level >= 4
        ? ["ERROR", "database connection acquisition timed out"]
        : level >= 3
          ? ["WARN", "database pool utilization above threshold"]
          : level >= 2
            ? ["WARN", "connection checkout duration increasing"]
            : [
                "INFO",
                level === 1
                  ? "order-service v2.14.7 deployment healthy"
                  : "request completed",
              ];
  return {
    id: `log-${state.tick}-${state.elapsedSeconds}`,
    second: state.elapsedSeconds,
    level: messages[0] as LogEntry["level"],
    service: level >= 2 ? "order" : "gateway",
    message: messages[1]!,
    fields: {
      request_id: `sim-${state.seed.toString(16)}-${state.tick}`,
      duration_ms: metricFor(state).orderLatencyMs,
    },
  };
}

const systemEvents: Record<
  IncidentStage,
  Omit<TimelineEntry, "id" | "second"> | undefined
> = {
  Normal: {
    kind: "system",
    title: "Incident simulation started",
    description: "Traffic and service health are within normal bounds.",
  },
  "Deployment completed": {
    kind: "system",
    title: "Deployment completed",
    description: "Order service version 2.14.7 finished a rolling deployment.",
  },
  "Connection leak begins": {
    kind: "system",
    title: "First latency increase",
    description: "Early order-service database wait time begins to rise.",
  },
  "Database pool saturation": {
    kind: "alert",
    title: "Database pool saturation",
    description:
      "Active orders database connections approach the configured ceiling.",
  },
  "Order-service latency increase": {
    kind: "alert",
    title: "Order-service latency alert",
    description: "The latency SLO burn alert is now active.",
  },
  "Checkout errors": {
    kind: "alert",
    title: "Checkout error-rate increase",
    description: "Customer checkout failures cross the critical threshold.",
  },
  "Incident mitigation": undefined,
  Recovery: {
    kind: "recovery",
    title: "Recovery signals detected",
    description: "Connection pressure and checkout latency are falling.",
  },
  Completed: {
    kind: "recovery",
    title: "Incident completed",
    description: "Simulated checkout health has returned to baseline.",
  },
};

function initialStateFromSeed(
  scenarioId: string,
  seed: number,
): SimulationState {
  const base: SimulationState = {
    scenarioId,
    seed,
    elapsedSeconds: 0,
    tick: 0,
    stage: "Normal",
    status: "ready",
    speed: 1,
    services: [],
    logs: [],
    metrics: [],
    timeline: [{ id: "timeline-start", second: 0, ...systemEvents.Normal! }],
    collectedEvidence: [],
    hypotheses: [],
    actions: [],
    notes: "",
    selectedServiceId: null,
    activeTool: "overview",
    mitigationAt: null,
    modifiers: {
      poolBonus: 0,
      latencyReduction: 0,
      errorReduction: 0,
      consumerPaused: false,
    },
    announcement: "",
  };
  return { ...base, services: servicesFor(base), metrics: [metricFor(base)] };
}
export function createInitialState(
  scenarioId: string,
  sessionId: string,
): SimulationState {
  return initialStateFromSeed(
    scenarioId,
    seedFromSession(scenarioId, sessionId),
  );
}

export function advanceSimulation(state: SimulationState): SimulationState {
  if (state.status === "completed") return state;
  const elapsedSeconds = state.elapsedSeconds + SIMULATION_INTERVAL_SECONDS;
  const tick = state.tick + 1;
  const stage = stageFor(elapsedSeconds, state.mitigationAt);
  const draft = { ...state, elapsedSeconds, tick, stage };
  const changed = stage !== state.stage;
  const event = changed ? systemEvents[stage] : undefined;
  const next: SimulationState = {
    ...draft,
    status: stage === "Completed" ? "completed" : state.status,
    services: servicesFor(draft),
    metrics: [...state.metrics, metricFor(draft)].slice(-MAX_METRICS),
    logs: [...state.logs, logFor(draft)].slice(-MAX_LOGS),
    timeline: event
      ? [
          ...state.timeline,
          { id: `timeline-${tick}-${stage}`, second: elapsedSeconds, ...event },
        ]
      : state.timeline,
    announcement: event?.kind === "alert" ? event.title : "",
  };
  return next;
}

function performAction(
  state: SimulationState,
  action: ActionId,
): SimulationState {
  const definition = actionDefinitions[action];
  const id = `action-${state.tick}-${state.actions.length + 1}`;
  let mitigationAt = state.mitigationAt;
  const modifiers = { ...state.modifiers };
  let effect =
    "Observation interval recorded; simulation state continues unchanged.";
  if (action === "rollback") {
    mitigationAt = state.elapsedSeconds;
    effect =
      "Rollback initiated; the simulation enters mitigation and recovery can begin.";
  }
  if (action === "increase-pool") {
    modifiers.poolBonus += 10;
    modifiers.latencyReduction += 300;
    effect =
      "Pool ceiling increases temporarily, reducing wait pressure without removing the underlying cause.";
  }
  if (action === "scale") {
    modifiers.latencyReduction += 180;
    effect = "Additional order-service capacity temporarily lowers latency.";
  }
  if (action === "disable-retry") {
    modifiers.errorReduction += 2;
    effect = "Retry amplification falls, reducing simulated error pressure.";
  }
  if (action === "pause-consumer") {
    modifiers.consumerPaused = true;
    effect =
      "Message consumption pauses and notification backlog begins to accumulate.";
  }
  if (action === "restart") {
    modifiers.latencyReduction += 120;
    effect =
      "An order-service instance restarts; symptoms ease briefly but may return.";
  }
  const entry = {
    id,
    action,
    label: definition.label,
    second: state.elapsedSeconds,
    risk: definition.risk,
    effect,
  };
  const stage =
    mitigationAt !== state.mitigationAt ? "Incident mitigation" : state.stage;
  return {
    ...state,
    mitigationAt,
    modifiers,
    stage,
    services: servicesFor({ ...state, modifiers, stage }),
    actions: [...state.actions, entry],
    timeline: [
      ...state.timeline,
      {
        id: `timeline-${id}`,
        second: state.elapsedSeconds,
        kind: "action",
        title: definition.label,
        description: effect,
      },
    ],
    announcement: `${definition.label} recorded. ${effect}`,
  };
}

export function simulationReducer(
  state: SimulationState,
  event: SimulationEvent,
): SimulationState {
  switch (event.type) {
    case "START":
      return state.status === "ready" ? { ...state, status: "running" } : state;
    case "PAUSE":
      return state.status === "running"
        ? { ...state, status: "paused" }
        : state;
    case "RESUME":
      return state.status === "paused"
        ? { ...state, status: "running" }
        : state;
    case "ADVANCE":
      return advanceSimulation(state);
    case "RESET":
      return initialStateFromSeed(state.scenarioId, state.seed);
    case "SET_SPEED":
      return { ...state, speed: event.speed };
    case "SELECT_SERVICE":
      return { ...state, selectedServiceId: event.serviceId };
    case "SET_TOOL":
      return { ...state, activeTool: event.tool };
    case "SET_NOTES":
      return { ...state, notes: event.notes };
    case "COLLECT_EVIDENCE":
      if (state.collectedEvidence.some(({ id }) => id === event.evidence.id))
        return state;
      return {
        ...state,
        collectedEvidence: [
          ...state.collectedEvidence,
          { ...event.evidence, annotation: "", hypothesisIds: [] },
        ],
        announcement: `Evidence collected from ${event.evidence.source}.`,
      };
    case "ANNOTATE_EVIDENCE":
      return {
        ...state,
        collectedEvidence: state.collectedEvidence.map((item) =>
          item.id === event.evidenceId
            ? { ...item, annotation: event.annotation }
            : item,
        ),
      };
    case "CREATE_HYPOTHESIS": {
      const hypothesis = {
        id: `hyp-${state.hypotheses.length + 1}`,
        title: event.title,
        notes: event.notes,
        status: "unresolved" as const,
        evidenceIds: [],
      };
      return { ...state, hypotheses: [...state.hypotheses, hypothesis] };
    }
    case "SET_HYPOTHESIS_STATUS":
      return {
        ...state,
        hypotheses: state.hypotheses.map((item) =>
          item.id === event.hypothesisId
            ? { ...item, status: event.status }
            : item,
        ),
      };
    case "ATTACH_EVIDENCE":
      return {
        ...state,
        hypotheses: state.hypotheses.map((item) =>
          item.id === event.hypothesisId &&
          !item.evidenceIds.includes(event.evidenceId)
            ? { ...item, evidenceIds: [...item.evidenceIds, event.evidenceId] }
            : item,
        ),
        collectedEvidence: state.collectedEvidence.map((item) =>
          item.id === event.evidenceId &&
          !item.hypothesisIds.includes(event.hypothesisId)
            ? {
                ...item,
                hypothesisIds: [...item.hypothesisIds, event.hypothesisId],
              }
            : item,
        ),
      };
    case "PERFORM_ACTION":
      return performAction(state, event.action);
  }
}

export function availableEvidence(state: SimulationState) {
  return evidenceCatalog.filter(
    (item) => item.availableAt <= state.elapsedSeconds,
  );
}
export function canRevealConclusion(state: SimulationState): boolean {
  const ids = new Set(state.collectedEvidence.map(({ id }) => id));
  return (
    state.stage === "Completed" ||
    [
      "ev-deploy-2147",
      "ev-metric-pool",
      "ev-log-timeout",
      "ev-trace-wait",
    ].every((id) => ids.has(id))
  );
}

export function playerVisibleStage(state: SimulationState): string {
  if (state.stage === "Connection leak begins" && !canRevealConclusion(state))
    return "Early anomaly developing";
  return state.stage;
}

File: services/api/app/simulation/engine.py
from __future__ import annotations

from copy import deepcopy
from datetime import UTC, datetime, timedelta
from typing import Any

from app.simulation.registry import ScenarioDefinition

INTERVAL_SECONDS = 30
MAX_LOGS = 100
MAX_METRICS = 120
MAX_TRACES = 60
MAX_EVENTS = 160
MAX_ACTIONS = 100
ENGINE_VERSION = "sentinelops-engine@1.0.0"


def _hash(seed: int, tick: int, salt: int) -> int:
    value = (seed ^ ((tick + 1) * 0x9E3779B1) ^ salt) & 0xFFFFFFFF
    value ^= (value << 13) & 0xFFFFFFFF
    value ^= value >> 17
    value ^= (value << 5) & 0xFFFFFFFF
    return value & 0xFFFFFFFF


def _hex(seed: int, tick: int, length: int, salt: int) -> str:
    parts: list[str] = []
    index = 0
    while len("".join(parts)) < length:
        parts.append(f"{_hash(seed, tick + index, salt + index):08x}")
        index += 1
    return "".join(parts)[:length]


def _timestamp(second: int) -> str:
    return (datetime(2026, 1, 15, 0, 0, tzinfo=UTC) + timedelta(seconds=second)).isoformat()


def stage_for(second: int, mitigation_at: int | None) -> str:
    if mitigation_at is not None:
        since = second - mitigation_at
        if since >= 120:
            return "Completed"
        if since >= 60:
            return "Recovery"
        return "Incident mitigation"
    if second >= 150:
        return "Checkout errors"
    if second >= 120:
        return "Order-service latency increase"
    if second >= 90:
        return "Database pool saturation"
    if second >= 60:
        return "Connection leak begins"
    if second >= 30:
        return "Deployment completed"
    return "Normal"


def _severity(stage: str) -> int:
    stages = [
        "Normal",
        "Deployment completed",
        "Connection leak begins",
        "Database pool saturation",
        "Order-service latency increase",
        "Checkout errors",
    ]
    return stages.index(stage) if stage in stages else 0


def initial_state(scenario: ScenarioDefinition, seed: int) -> dict[str, Any]:
    return {
        "scenarioId": scenario.public.id,
        "scenarioSlug": scenario.public.slug,
        "seed": seed,
        "elapsedSeconds": 0,
        "tick": 0,
        "stage": "Normal",
        "status": "ready",
        "speed": 1,
        "logs": [],
        "metrics": [_metric(seed, 0, 0, "Normal", {}, [])],
        "traces": [],
        "alerts": _alerts(),
        "timeline": [
            {
                "id": "timeline-start",
                "second": 0,
                "kind": "system",
                "title": "Incident simulation started",
                "description": "Traffic and service health are within normal bounds.",
            }
        ],
        "collectedEvidence": [],
        "hypotheses": [],
        "actions": [],
        "rootCauseSubmission": None,
        "recoveryVerification": None,
        "completionDocumentation": None,
        "report": None,
        "investigationCompleted": False,
        "mitigationAt": None,
        "modifiers": {"poolBonus": 0, "latencyReduction": 0, "errorReduction": 0},
    }


def _metric(
    seed: int,
    tick: int,
    second: int,
    stage: str,
    modifiers: dict[str, int],
    actions: list[dict[str, Any]],
) -> dict[str, Any]:
    level = _severity(stage)
    jitter = _hash(seed, tick, 99) % 23
    restart_times = [item["second"] for item in actions if item["action"] == "restart"]
    restart_relief = 120 if restart_times and second - max(restart_times) <= 60 else 0
    recovery = 0 if stage == "Completed" else 0.35 if stage == "Recovery" else 1
    pool_max = 40 + modifiers.get("poolBonus", 0)
    pool_used = min(pool_max, round((8 + level * 7) * recovery))
    return {
        "second": second,
        "requestRate": 780 + jitter,
        "orderLatencyMs": max(
            80,
            round(
                (95 + level * 480 - modifiers.get("latencyReduction", 0) - restart_relief)
                * recovery
                + jitter
            ),
        ),
        "latencyP50Ms": max(45, round((70 + level * 85) * recovery + jitter / 2)),
        "latencyP99Ms": max(110, round((130 + level * 620) * recovery + jitter)),
        "checkoutErrorRate": round(
            max(0.1, (12.8 if level >= 5 else 2.2 if level >= 4 else 0.2) * recovery), 1
        ),
        "cpuPercent": round(min(92, 31 + level * 7 + jitter / 5), 1),
        "memoryMb": 420 + level * 18 + jitter,
        "dbPoolUsed": pool_used,
        "dbPoolMax": pool_max,
        "dbPoolUtilizationPercent": round(pool_used / pool_max * 100),
        "queueDepth": max(2, 12 + level * 2),
        "serviceRestarts": sum(item["action"] == "restart" for item in actions),
    }


def _logs(
    seed: int, tick: int, second: int, stage: str, metric: dict[str, Any]
) -> list[dict[str, Any]]:
    level = _severity(stage)
    severity, message = (
        ("ERROR", "database connection acquisition timed out")
        if level >= 4
        else ("WARN", "database pool utilization above threshold")
        if level >= 3
        else ("WARN", "connection checkout duration increasing")
        if level >= 2
        else ("INFO", "order-service deployment healthy")
        if level == 1
        else ("INFO", "request completed")
    )
    trace_id = _hex(seed, tick, 32, 17)
    request_id = f"req_sim_{_hex(seed, tick, 12, 8)}"
    base = {
        "second": second,
        "timestamp": _timestamp(second),
        "traceId": trace_id,
        "deploymentVersion": "2.14.7" if level >= 1 else "2.14.6",
        "requestId": request_id,
    }
    return [
        {
            **base,
            "id": f"log-{tick}-order",
            "level": severity,
            "service": "order",
            "serviceName": "order-service",
            "spanId": _hex(seed, tick, 16, 3),
            "message": message,
            "fields": {
                "service.name": "order-service",
                "deployment.version": base["deploymentVersion"],
                "request_id": request_id,
                "duration_ms": metric["orderLatencyMs"],
                "sentinelops.simulated": True,
            },
        },
        {
            **base,
            "id": f"log-{tick}-gateway",
            "level": "ERROR" if level >= 5 else "INFO",
            "service": "gateway",
            "serviceName": "api-gateway",
            "spanId": _hex(seed, tick, 16, 1),
            "message": "checkout response returned upstream error"
            if level >= 5
            else "checkout request routed",
            "fields": {
                "service.name": "api-gateway",
                "http.route": "/checkout",
                "sentinelops.simulated": True,
            },
        },
    ]


def _trace(seed: int, tick: int, second: int, metric: dict[str, Any]) -> dict[str, Any]:
    trace_id = _hex(seed, tick, 32, 17)
    duration = metric["orderLatencyMs"] + 46
    spans = [
        (1, None, "POST /checkout", "gateway", duration),
        (2, 1, "authorize", "auth", 24),
        (3, 1, "create order", "order", metric["orderLatencyMs"]),
        (4, 3, "acquire database connection", "orders-db", max(18, metric["orderLatencyMs"] - 95)),
        (5, 3, "INSERT orders", "orders-db", 62),
    ]
    return {
        "id": trace_id,
        "second": second,
        "timestamp": _timestamp(second),
        "rootService": "gateway",
        "durationMs": duration,
        "status": "ERROR" if metric["checkoutErrorRate"] >= 5 else "OK",
        "spans": [
            {
                "id": _hex(seed, tick, 16, index),
                "parentId": _hex(seed, tick, 16, parent) if parent else None,
                "name": name,
                "service": service,
                "startMs": index * 4,
                "durationMs": span_duration,
                "status": "ERROR" if index == 4 and metric["checkoutErrorRate"] >= 5 else "OK",
                "attributes": {"sentinelops.simulated": True, "db.system": "postgresql"}
                if index >= 4
                else {"sentinelops.simulated": True},
                "critical": index in {1, 3, 4},
                "relatedLogIds": [f"log-{tick}-order"] if index in {3, 4} else [],
            }
            for index, parent, name, service, span_duration in spans
        ],
    }


def _alerts() -> list[dict[str, Any]]:
    return [
        {
            "id": "alert-db-pool",
            "title": "Orders database pool saturation",
            "severity": "warning",
            "source": "metrics",
            "service": "orders-db",
            "firstTriggered": 90,
            "lastUpdated": 90,
            "status": "firing",
            "assignedTo": None,
        },
        {
            "id": "alert-order-latency",
            "title": "Order-service latency SLO burn",
            "severity": "critical",
            "source": "metrics",
            "service": "order",
            "firstTriggered": 120,
            "lastUpdated": 120,
            "status": "firing",
            "assignedTo": None,
        },
        {
            "id": "alert-checkout-errors",
            "title": "Checkout error rate elevated",
            "severity": "critical",
            "source": "logs",
            "service": "gateway",
            "firstTriggered": 150,
            "lastUpdated": 150,
            "status": "firing",
            "assignedTo": None,
        },
    ]


def advance(state: dict[str, Any]) -> dict[str, Any]:
    if state["status"] == "completed":
        return deepcopy(state)
    result = deepcopy(state)
    result["elapsedSeconds"] += INTERVAL_SECONDS
    result["tick"] += 1
    previous_stage = result["stage"]
    result["stage"] = stage_for(result["elapsedSeconds"], result["mitigationAt"])
    if result["stage"] == "Completed":
        result["status"] = "completed"
    metric = _metric(
        result["seed"],
        result["tick"],
        result["elapsedSeconds"],
        result["stage"],
        result["modifiers"],
        result["actions"],
    )
    logs = _logs(result["seed"], result["tick"], result["elapsedSeconds"], result["stage"], metric)
    result["metrics"] = (result["metrics"] + [metric])[-MAX_METRICS:]
    result["logs"] = (result["logs"] + logs)[-MAX_LOGS:]
    result["traces"] = (
        result["traces"]
        + [_trace(result["seed"], result["tick"], result["elapsedSeconds"], metric)]
    )[-MAX_TRACES:]
    for alert in result["alerts"]:
        if alert["firstTriggered"] <= result["elapsedSeconds"]:
            alert["lastUpdated"] = result["elapsedSeconds"]
    if result["stage"] != previous_stage:
        result["timeline"].append(
            {
                "id": f"timeline-{result['tick']}-{result['stage']}",
                "second": result["elapsedSeconds"],
                "kind": "recovery" if result["stage"] in {"Recovery", "Completed"} else "alert",
                "title": result["stage"],
                "description": "Deterministic simulated incident state updated.",
            }
        )
        result["timeline"] = result["timeline"][-MAX_EVENTS:]
    return result


def perform_action(
    state: dict[str, Any], action: str, target_id: str | None = None
) -> dict[str, Any]:
    result = deepcopy(state)
    if len(result["actions"]) >= MAX_ACTIONS:
        raise ValueError("Action limit reached.")
    labels = {
        "restart": "Restart service",
        "scale": "Scale service",
        "rollback": "Roll back deployment",
        "increase-pool": "Increase database pool temporarily",
        "disable-retry": "Disable retry behavior",
        "pause-consumer": "Pause message consumer",
        "observe": "Do nothing and observe",
        "ack-alert": "Acknowledge alert",
        "assign-alert": "Assign alert to self",
        "silence-alert": "Silence alert in simulation",
    }
    effect = "Observation interval recorded; simulation state continues unchanged."
    if action in {"ack-alert", "assign-alert", "silence-alert"}:
        alert = next((item for item in result["alerts"] if item["id"] == target_id), None)
        if alert is None or alert["firstTriggered"] > result["elapsedSeconds"]:
            raise ValueError("Alert is not active.")
        if action == "ack-alert":
            alert["status"] = "acknowledged"
            effect = "Alert acknowledged; underlying telemetry remains active."
        elif action == "assign-alert":
            alert["assignedTo"] = "self"
            effect = "Alert assigned locally to the player."
        else:
            alert["status"] = "silenced"
            effect = "Alert notifications silenced; evidence and telemetry are retained."
    elif action == "rollback":
        if result["mitigationAt"] is not None:
            raise ValueError("Rollback has already been applied.")
        result["mitigationAt"] = result["elapsedSeconds"]
        result["stage"] = "Incident mitigation"
        effect = "Rollback initiated; the simulation enters mitigation and recovery can begin."
    elif action == "increase-pool":
        result["modifiers"]["poolBonus"] += 10
        result["modifiers"]["latencyReduction"] += 300
        effect = "Pool ceiling increases temporarily, reducing simulated wait pressure."
    elif action == "scale":
        result["modifiers"]["latencyReduction"] += 180
        effect = "Additional order-service capacity temporarily lowers latency."
    elif action == "restart":
        effect = "An order-service instance restarts; symptoms ease briefly."
    elif action == "disable-retry":
        result["modifiers"]["errorReduction"] += 2
        effect = "Retry amplification falls, reducing simulated error pressure."
    entry = {
        "id": f"action-{result['tick']}-{len(result['actions']) + 1}",
        "action": action,
        "targetId": target_id,
        "label": labels[action],
        "second": result["elapsedSeconds"],
        "risk": "Safe simulated action; no real command is executed.",
        "effect": effect,
    }
    result["actions"].append(entry)
    result["timeline"] = (
        result["timeline"]
        + [
            {
                "id": f"timeline-{entry['id']}",
                "second": result["elapsedSeconds"],
                "kind": "action",
                "title": labels[action],
                "description": effect,
            }
        ]
    )[-MAX_EVENTS:]
    return result
